sdk/src/trace/tracer.cc

```diff
@@ -38,14 +40,19 @@ nostd::unique_ptr<trace_api::Span> Tracer::StartSpan(
       sampler_->ShouldSample(nullptr, trace_api::TraceId(), name, options.kind, attributes);
   if (sampling_result.decision == Decision::NOT_RECORD)
   {
-    return nostd::unique_ptr<trace_api::Span>{new (std::nothrow)
-                                                  trace_api::NoopSpan{this->shared_from_this()}};
+    auto span = nostd::shared_ptr<trace_api::Span>{
+        new (std::nothrow) trace_api::NoopSpan{this->shared_from_this()}};
+
+    return span;
   }
   else
   {
-    auto span = nostd::unique_ptr<trace_api::Span>{new (std::nothrow) Span{
+    auto span = nostd::shared_ptr<trace_api::Span>{new (std::nothrow) Span{
         this->shared_from_this(), processor_.load(), name, attributes, options}};
 
+    span->SetToken(new context::Token(context::RuntimeContext::Attach(
+        context::RuntimeContext::GetCurrent().SetValue(SPAN_KEY, span))));
```

Comment:
I'm wondering: Could we go with just storing the current `SpanContext` in the runtime context? Or do we need to store the whole span?

Reply:
When just storing the `SpanContext`, we could keep handling spans as unique pointers. And we would still have all the information we need to parent spans.

Reply:
That would simplify things.

Reply:
For the sake of keeping it consistent though, I think we have to go with the span. Reiley said that we need to have span stored in context.

Reply:
I see that other languages store the span, but only use the context, so I was wondering about that.

I'm poking a bit at the `nostd::unique_ptr` to `nostd::shared_ptr` change, which I think is caused buy the need to store the span in the context. I think this is critical.

With a unique pointer, a span gets automatically ended and destroyed when the span goes out of context:
```c++
{
  auto span = tracer->StartSpan("span");
}
```
At the end of the scope the unique pointer will go out of scope, the destructor will be invoked which calls `End`.

When returning a shared pointer and keeping another shared pointer copy in the context, this will not work anymore (when the shared pointer returned by `StartSpan` will go out of scope, there will still be the shared pointer copy in the context stored, thus preventing the destructor to be invoked. We would effectively force a call to `End`, which is optional now:
```c++
{
  auto span = tracer->StartSpan("span");
  span->End();
}
```

I'd voice a strong preference for keeping the ability to automatically end spans when they go out of scope.

Reply:
@reyang

Reply:
Reiley, do you have any thoughts?

Reply:
I think the reason span is a context slot, is that we might need to get the current active span from the context and close it. If we just put SpanContext, there is no way we can close the span.

e.g.
```csharp
void AfterReceiveRequest()
{
    var span = tracer.StartSpan("A");
}

void BeforeSendResponse()
{
    var span = tracer.GetCurrentSpan();
    span.attributes["http.status_code"] = 3;
    span.End();
}
```

Reply:
I see. The [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#tracer-operations) defines the ability to retrieve the current span as a SHOULD.

This is then a decision of trade-offs that we need to make: either we support the lifetime management of spans via scopes, or we support `GetCurrentSpan`.